File: Alcoholtracker/Models/StatusSkin.swift
import Foundation

// MARK: - StatusSkin
//
// Swappable text vocabulary for BACStatus level labels.
// Raw values are English for SwiftData backward compatibility.

enum StatusSkin: String, Codable, CaseIterable {
    case standard = "standard"
    case normal   = "normal"   // colloquial German (mit Felipe)
    case youth    = "youth"    // natural youth language, not forced slang
    case chill    = "chill"
    case sailor   = "sailor"
    case formal   = "formal"
    case science  = "science"
    case festival = "festival"
    case medieval = "medieval"
    case emoji    = "emoji"

    var displayName: String {
        switch self {
        case .standard: return "Standard"
        case .normal:   return "Normal"
        case .youth:    return "Alltag"
        case .chill:    return "Chill"
        case .sailor:   return "Seemann"
        case .formal:   return "Formal"
        case .science:  return "Wissenschaft"
        case .festival: return "Festival"
        case .medieval: return "Mittelalter"
        case .emoji:    return "Emoji"
        }
    }

    var skinDescription: String {
        switch self {
        case .standard: return "Klassische deutsche Bezeichnungen."
        case .normal:   return "So wie man wirklich redet."
        case .youth:    return "Alltägliche Sprache, nichts Aufgesetztes."
        case .chill:    return "Entspannte, positive Formulierungen."
        case .sailor:   return "Seemannsprache auf hoher See."
        case .formal:   return "Sachliche, medizinisch angelehnte Begriffe."
        case .science:  return "Physiologische Effektbeschreibungen."
        case .festival: return "Vom Warm Up bis zum Heimweg."
        case .medieval: return "Altdeutsche Trinkbegriffe."
        case .emoji:    return "Mit Farbkreisen auf einen Blick."
        }
    }

    var previewLabel: String { label(for: .tipsy) }

    func label(for status: BACStatus) -> String {
        switch self {
        case .standard:
            switch status {
            case .sober:   return "Nüchtern"
            case .tipsy:   return "Leicht beschwipst"
            case .drunk:   return "Beschwipst"
            case .careful: return "Aufpassen"
            case .danger:  return "Fahruntauglich"
            }
        case .normal:
            // User-requested names, exactly as specified
            switch status {
            case .sober:   return "Garnicht Drunk"
            case .tipsy:   return "Minimal Drunk"
            case .drunk:   return "Bisschen Drunk"
            case .careful: return "Komplett Drunk"
            case .danger:  return "Felipe"
            }
        case .youth:
            // Natural everyday German youth language, not cringe
            switch status {
            case .sober:   return "Clean"
            case .tipsy:   return "Leicht angeheitert"
            case .drunk:   return "Angemacht"
            case .careful: return "Durch den Wind"
            case .danger:  return "Komplett weg"
            }
        case .chill:
            switch status {
            case .sober:   return "Fit"
            case .tipsy:   return "Leicht angeheitert"
            case .drunk:   return "Gut drauf"
            case .careful: return "Langsam machen"
            case .danger:  return "Zu viel"
            }
        case .sailor:
            switch status {
            case .sober:   return "Klar Schiff"
            case .tipsy:   return "Leichter Seegang"
            case .drunk:   return "Volle Fahrt"
            case .careful: return "Stürmisch"
            case .danger:  return "Mann über Bord"
            }
        case .formal:
            switch status {
            case .sober:   return "Alkoholfrei"
            case .tipsy:   return "Gering"
            case .drunk:   return "Spürbar"
            case .careful: return "Erheblich"
            case .danger:  return "Gefährlich"
            }
        case .science:
            switch status {
            case .sober:   return "Basislinie"
            case .tipsy:   return "Schwellenwert"
            case .drunk:   return "Euphorisch"
            case .careful: return "Sediert"
            case .danger:  return "Toxisch"
            }
        case .festival:
            switch status {
            case .sober:   return "Ankunft"
            case .tipsy:   return "Warm Up"
            case .drunk:   return "Im Flow"
            case .careful: return "Volle Power"
            case .danger:  return "Nach Hause"
            }
        case .medieval:
            switch status {
            case .sober:   return "Klar"
            case .tipsy:   return "Heiter"
            case .drunk:   return "Fröhlich"
            case .careful: return "Angetrunken"
            case .danger:  return "Besoffen"
            }
        case .emoji:
            switch status {
            case .sober:   return "Fit 🟢"
            case .tipsy:   return "Happy 🟡"
            case .drunk:   return "Woah 🟠"
            case .careful: return "Viel 🔴"
            case .danger:  return "Stop 🚫"
            }
        }
    }
}

File: android/bac/src/main/kotlin/de/tipau/promille/bac/StatusSkin.kt
package de.tipau.promille.bac

/**
 * Which band a permille value falls into. Mirrors BACStatus in
 * Models/DrinkTemplate.swift; [level] is the ordinal iOS exposes.
 */
enum class BacStatus(val level: Int, val germanName: String) {
    SOBER(0, "Nüchtern"),
    TIPSY(1, "Leicht beschwipst"),
    DRUNK(2, "Beschwipst"),
    // 0,8 to 1,5 permille is already well over the 0,5 legal limit, so this band
    // says "unfit to drive" rather than a softer "watch out".
    CAREFUL(3, "Fahruntüchtig"),
    DANGER(4, "Gefährlich");

    fun label(skin: StatusSkin): String = skin.label(this)

    companion object {
        /** Fixed default bands, used when no profile is loaded yet. */
        fun of(bac: Double): BacStatus = when {
            bac < 0.01 -> SOBER
            bac < 0.3 -> TIPSY
            bac < 0.8 -> DRUNK
            bac < 1.5 -> CAREFUL
            else -> DANGER
        }

        /** User-adjustable bands. A null profile falls back to [of]. */
        fun of(bac: Double, profile: Profile?): BacStatus {
            val p = profile ?: return of(bac)
            return when {
                bac < p.tipsyThreshold -> SOBER
                bac < p.drunkThreshold -> TIPSY
                bac < p.carefulThreshold -> DRUNK
                bac < p.dangerThreshold -> CAREFUL
                else -> DANGER
            }
        }
    }
}

/**
 * Swappable text vocabulary for the status labels. Raw values are English for
 * SwiftData backward compatibility and cross-platform sync; do not translate them.
 */
enum class StatusSkin(
    val raw: String,
    val displayName: String,
    val skinDescription: String
) {
    STANDARD("standard", "Standard", "Klassische deutsche Bezeichnungen."),
    NORMAL("normal", "Normal", "So wie man wirklich redet."),
    YOUTH("youth", "Alltag", "Alltägliche Sprache, nichts Aufgesetztes."),
    CHILL("chill", "Chill", "Entspannte, positive Formulierungen."),
    SAILOR("sailor", "Seemann", "Seemannsprache auf hoher See."),
    FORMAL("formal", "Formal", "Sachliche, medizinisch angelehnte Begriffe."),
    SCIENCE("science", "Wissenschaft", "Physiologische Effektbeschreibungen."),
    FESTIVAL("festival", "Festival", "Vom Warm Up bis zum Heimweg."),
    MEDIEVAL("medieval", "Mittelalter", "Altdeutsche Trinkbegriffe."),
    EMOJI("emoji", "Emoji", "Mit Farbkreisen auf einen Blick.");

    /** What the picker shows as a sample: always the TIPSY band. */
    val previewLabel: String get() = label(BacStatus.TIPSY)

    fun label(status: BacStatus): String {
        val i = status.level
        return LABELS.getValue(this)[i]
    }

    companion object {
        fun from(raw: String): StatusSkin = entries.firstOrNull { it.raw == raw } ?: STANDARD

        // Order per row is sober, tipsy, drunk, careful, danger.
        private val LABELS: Map<StatusSkin, List<String>> = mapOf(
            STANDARD to listOf(
                "Nüchtern", "Leicht beschwipst", "Beschwipst", "Fahruntüchtig", "Gefährlich"
            ),
            // User-requested names, exactly as specified.
            NORMAL to listOf(
                "Garnicht Drunk", "Minimal Drunk", "Bisschen Drunk", "Komplett Drunk", "Felipe"
            ),
            // Natural everyday German youth language, not cringe.
            YOUTH to listOf(
                "Clean", "Leicht angeheitert", "Angemacht", "Durch den Wind", "Komplett weg"
            ),
            CHILL to listOf(
                "Fit", "Leicht angeheitert", "Gut drauf", "Langsam machen", "Zu viel"
            ),
            SAILOR to listOf(
                "Klar Schiff", "Leichter Seegang", "Volle Fahrt", "Stürmisch", "Mann über Bord"
            ),
            FORMAL to listOf(
                "Alkoholfrei", "Gering", "Spürbar", "Erheblich", "Gefährlich"
            ),
            SCIENCE to listOf(
                "Basislinie", "Schwellenwert", "Euphorisch", "Sediert", "Toxisch"
            ),
            FESTIVAL to listOf(
                "Ankunft", "Warm Up", "Im Flow", "Volle Power", "Nach Hause"
            ),
            MEDIEVAL to listOf(
                "Klar", "Heiter", "Fröhlich", "Angetrunken", "Besoffen"
            ),
            EMOJI to listOf(
                "Fit 🟢", "Happy 🟡", "Woah 🟠", "Viel 🔴", "Stop 🚫"
            )
        )
    }
}
